Switch `cast` userscript from youtube-dl to yt-dlp

DavidRV00 opened the pull request:
The `cast` userscript hasn't worked for me in a while (qutebrowser says "Casting \<my url\>" but the TV just shows black), because it attempts to launch youtube-dl which has been replaced by yt-dlp for some time. Switching this one command gets `cast` working for me.

<!-- Thanks for submitting a pull request! Please pick a descriptive title (not just "issue 12345"). If there is an open issue associated to your PR, please add a line like "Closes #12345" somewhere in the PR description (outside of this comment) -->

mschilli87 commented:
While I personally use `yt-dlp` myself as well, AFAIK, `youtube-dl` is not dead as a project. Also, the same effect could be achieved by creating a `youtube-dl` symlink/wrapper anywhere on `$PATH`. Finally, if we are already touching this script, why not make the download tool name configurable, no matter if we leave the default at `youtube-dl` or change it to `yt-dlp`? That way we'd be already solving the next iteration of this problem alongside the current one. :wink:

DavidRV00 commented:
Sure, thanks for the response. I'm almost done with a revision that adds configurability. Just testing it now and then I'll send it to see what you think.

mschilli87 commented on misc/userscripts/cast:
```diff
@@ -133,23 +144,40 @@ echo "jseval -q $(printjs)" >> "$QUTE_FIFO"
     exit 1
 fi
+if [[ "${ytdl_program}" == "" ]]; then
+    msg error "youtube-dl (or a drop-in replacement) can't be found..."
```
Thanks for the improvements. One nitpick: IMHO it would be nice to include the failing configuration variable value in the error message:

```suggestion
    msg error "youtube-dl (or a drop-in replacement) '${ytdl_program}' can't be found..."
```

The-Compiler replied:
That would always be empty here, no? Not sure if it makes sense to complicate the logic just for this, though.

mschilli87 replied:
Sorry, I didn't read the actual condition. In this case I find the erro message misleading: If I see an error

> `"youtube-dl (or a drop-in replacement) can't be found..."`

I'd assume I the configured program was not found on my system, not that no program had been configured in the first place.
So how about:

```suggestion
    msg error "No youtube-dl (or a drop-in replacement) program configured..."
```

mschilli87 replied:
After looking more closely at the code again ([line159](https://github.com/qutebrowser/qutebrowser/pull/7074/files#diff-34b4768e7037f827d7c3cf3006f417c522b580730cdc1e94506f202c87f7ac83R159)) I think I *finally* get the logic. So I guess the meesage is correct and there is no way to check anymore if ${tydl_program}` was specified at this point. Maybe when switching to an ENVIRONMENT variable, it could be checked and included here, if defined.
Sorry for the noise.

emanuele6 commented on misc/userscripts/cast:
```diff
@@ -133,23 +144,41 @@ echo "jseval -q $(printjs)" >> "$QUTE_FIFO"
 
 # kill any running instance of castnow
-pkill -f "${program_}"
+pkill -f "${cast_program}"
```
```suggestion
pkill -f -- "${cast_program}"
```

The-Compiler commented:
FWIW I agree this might getting a bit overboard with the config file and such, especially since I don't think we'll have many more options for this script. An environment variable or perhaps commandline argument (though parsing those is some complexity too) might indeed be more appropriate.

emanuele6 commented on misc/userscripts/cast:
```diff
@@ -133,23 +144,41 @@ echo "jseval -q $(printjs)" >> "$QUTE_FIFO"
 
 # start youtube download in stream mode (-o -) into temporary file
-youtube-dl -qo - "$1" > "${file_to_cast}" &
+${ytdl_program} -qo - "$1" > "${file_to_cast}" &
```
Since `$ytdl_program`, at this point, is the result of `$(command -v "$p")` (see [line159](https://github.com/qutebrowser/qutebrowser/pull/7074/files#diff-34b4768e7037f827d7c3cf3006f417c522b580730cdc1e94506f202c87f7ac83R159)), this expansion should be quoted and it is incorrect to not quote it.

```suggestion
"${ytdl_program}" -qo - "$1" > "${file_to_cast}" &
```

DavidRV00 commented:
Thanks for the feedback everyone. I'll take a look and get something out when I can, maybe tomorrow or the next day.

DavidRV00 commented:
Switched it to just be an env var, and fixed the other things you all mentioned.

Made the error message more verbose. I could personally take-or-leave it.

The-Compiler commented:
Thanks @DavidRV00, @mschilli87 and @emanuele6! The shellcheck failure seems like a false-positive, I reported it at https://github.com/koalaman/shellcheck/issues/2479. In f654c0fdfc21ebf7fc8f601a6800314bb570f226 I changed the code slightly to avoid the issue.

There seems to be another issue with qutebrowser interpreting the `'` in `can't` despite it being quoted, but I don't really understand what's going on there... I opened #7089 to track that.
